tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -1151,6 +1151,46 @@ fillSymbolInfo(CursorSymbolInfo &Symbol, const DeclInfo &DInfo,
   return llvm::Error::success();
 }
 
+static bool addCursorInfoForLiteral(
+    CursorInfoData &Data, Expr *LitExpr, SwiftLangSupport &Lang,
+    const CompilerInvocation &CompInvoc, SourceLoc CursorLoc,
+    ArrayRef<ImmutableTextSnapshotRef> PreviousSnaps) {
+  if (!LitExpr) {
+    return false;
+  }
+
+  // Tries to retrieve the referenced initializer.
+  auto getInitializer = [](auto *Expr) -> ValueDecl * {
+    if (auto *LitExpr = dyn_cast<LiteralExpr>(Expr)) {
+      if (LitExpr->getInitializer().getDecl()) {
+        return LitExpr->getInitializer().getDecl();
+      }
+      if (auto *BuiltinLit = dyn_cast<BuiltinLiteralExpr>(LitExpr)) {
+        return BuiltinLit->getBuiltinInitializer().getDecl();
```

Comment:
To get the decl of the type we are constructing, I would do the following. AFAICT, this requires to make `getInitializer` a static function instead of a lambda. And I would rename it to `getCursorInfoDeclForLiteral` + change the comment to `If `Expr` is a literal, returns the declaration that should be reported by cursor info for that initializer.`

```suggestion
      // We shouldn’t report the builtin initializer to the user because it’s underscored and not visible.
      // Instead, return the type of the literal instead.
      if (auto Ty = LitExpr->getType()) {
        if (auto NominalTy = Ty->getAs<NominalOrBoundGenericNominalType>()) {
          return NominalTy->getDecl();
        }
      }
```

Reply:
I believe it still compiles fine in a lambda but I moved it into the static function since the logic became a bit more complicated 🙂 I structured this whole function a bit differently to avoid that many nested blocks.